### ISSUE TITLE
Investigate the use of a Promises based interface

### ISSUE THREAD

**Issue by angrykoala** [angrykoala/yerbamate#47]
it could work along with the current callback system. Related to #30

**Comment by ncuillery** [comment]
Hi, I just came across this repo via the Hacktoberfest label.

If I understand well, you wish that the `run` command would work either with a callback in last argument (returns `undefined`) or without a callback (returns a promise)?

```javascript
run('ls -l', (code, stdout) => console.log('Result:', code, stdout)) // callback
run('ls -l')
  .then(({code, stdout}) => console.log('Result:', code, stdout)) // promise
```

Am I right ?

If so, this pattern can be easily implemented with modules like [pify](https://www.npmjs.com/package/pify), [thenify](https://www.npmjs.com/package/thenify), [es6-promisify](https://www.npmjs.com/package/es6-promisify) as well as the [promisify](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original) method of Node.js `util` package (since v8.0.0).

The problem is: the `run` method doesn't follow the "error first" callback convention:

No matter what happen (errors or not), the callback arguments are:
1. exit code
2. out
3. errs

To stick to the Node convention and make yerbamate easier to use with Promise (or [async](https://github.com/caolan/async)), it should be:

In case of success (exit code === 0):
1. `null`
2. out
3. errs

In case of failure:
1. An instance of `Error` with the exit code
2. out
3. errs

So I'd recommend to do that instead of handle both callback and promise. What do you think ?

**Comment by angrykoala** [comment]
Thanks for your ideas @ncuillery 

First, I wouldn't like to add unnecessary libraries or use Promisify, I don't thinks is worth it as native Promises are simple enough.

I agree that the interface is not following the Node convention, on the other hand I'm not sure it makes sense to follow it, as the exit code seems more reasonable to return in this particular case than `undefined` or `Error`. Keep in mind that the _yerbamate_ operations will not "fail" as a failed execution is part of the expected behaviour (in fact, because is test-oriented, in some cases an execution is expected to return a different code).

It is also true, that under these conditions, a Promise-based interface may not be the most logical thing to do, as it will either always resolve succesfully, or reject when code!==0 (in the same way you proposed)

Thoughts on this @ncuillery ?

**Comment by ncuillery** [comment]
>First, I wouldn't like to add unnecessary libraries or use Promisify, I don't thinks is worth it as native Promises are simple enough.

Well, _promisify_ is not a full Promise implementation, it is only a utility function that tranforms a callback-based API into a promise-based API. 

The source code probably looks like this:

```javascript
function promisify(callbackBasedFn) {
  return function promiseBasedFn() {
    const args = arguments;
    return new Promise((resolve, reject) => {
      callbackBasedFn(...args, (err, result) => {
        err ? reject(err) : resolve(result);
      });
    });
  }
}
```

My thought was to allow users to promisify the `run` command on demand, not to include promisify. 

>I agree that the interface is not following the Node convention, on the other hand I'm not sure it makes sense to follow it, as the exit code seems more reasonable to return in this particular case than `undefined` or `Error`. 

The problem with the "error first" callback convention is that the users may write this code:

```javascript
run('ls -l', (err, result) => {
  if (err) {
    console.log('Ooops, something went wrong');
  } else {
    console.log('Everything is fine')
  }
})
```

Since the exit code is the first argument and its value in case of success (`0`) is falsy in JavaScript, the behavior is the opposite of the expected one 😜 

>Keep in mind that the yerbamate operations will not "fail" as a failed execution is part of the expected behaviour (in fact, because is test-oriented, in some cases an execution is expected to return a different code).

Why not ? 😇 Testing error and rejection is a common thing, with chai-as-promised, it could be done with clean expectations:

```javascript
run("node " + testScript, testDir, { args: "-error" })
  .should.be.rejectedWith(Error).with.property('code', 1);
```

>It is also true, that under these conditions, a Promise-based interface may not be the most logical thing to do, as it will either always resolve succesfully, or reject when code!==0 (in the same way you proposed)

I agree with this. From my experience, testing a promise-based API could be tricky, even with chai-as-promised, I don't think the tests will be cleaner or less verbose than before.

Promises are useful in case of multiple calls, sequential (especially when the arguments of a call depend on the result of the previous one) or simultaneous (with `Promise.all`). 

Does it correspond to some _yerbamate_ use cases ?

If not, well, maybe promise support won't be such a big improvement 😎

**Comment by angrykoala** [comment]
Good points.
I misunderstood about the promisify,  I agree with most of them, my only question then, is how an error case would be tested under this new interface? e.g I want to check the returned code is 2.

While seems logical to make the interface a bit more node-convention friendly, actually using node's promisify requires only two arguments (err,result), adding all the result into a single object may be confusing (or not?)

About the last point, while yerbamate is tests-oriented, an use case in https://github.com/angrykoala/gaucho uses yerbamate not for testing, but for actual execution of scripts, this could benefit from promises interfaces

**Comment by ncuillery** [comment]
>how an error case would be tested under this new interface? e.g I want to check the returned code is 2

Here is the snippet I posted above. I guess it should do the trick:

```javascript
run("node " + testScript, testDir, { args: "-error" })
  .should.be.rejectedWith(Error).with.property('code', 2);
```

>While seems logical to make the interface a bit more node-convention friendly, actually using node's promisify requires only two arguments (err,result), adding all the result into a single object may be confusing (or not?)

Yes, that's the limit of the Promise API (either built by hand or using promisify): you can only resolve or reject one value. But you can keep multiple args in callback signature (`out, errs`), `util.promisify` handles multi-args callbacks by reducing additional arguments into a single object.

Some of the Node.js API requires that feature. See: http://2ality.com/2017/05/util-promisify.html#promisifying-functions-whose-callbacks-have-more-than-two-parameters. I wouldn't say it's "confusing": it's true that it is something that users have to learn. Moreover, the way `promisify` get the names of the callback arguments is very 💩
 (https://stackoverflow.com/questions/1007981/how-to-get-function-parameter-names-values-dynamically). But, IMO it doesn't matter, that how these `promisify` libs works internally. The most important thing here is to stick to the convention, then, people are able to use `promisify` or not.

To summarize, 2 options:
- Keep the callback API like it is and add a custom Promise support that always resolve.
- Make the callback API following the "error-first" convention so that it could be easily wrapped by `promisify`.

It is totally up to you, it's true that the latter introduces a breaking change.

**Comment by angrykoala** [comment]
Ok, @ncuillery What do you think of the following?

1. The interface will return a callback with 2 arguments:
  * `err` or `null`, err will be an instance or `Error` with the result code as message. If the result code is 0, the first argument is null/undefined (whatever you consider better)

  * `result` will be an object with 3 variables:
     * `stdout`: same as the current `out`
     * `stderr`: same as current `err`
     * `code`: same as current `code`

2. If no callback is specified, a promise will be returned, the promise will be rejected if code!==0, the status code will be returned. if code==0, the same result as before will be returned as "resolved"

Any thoughts on this?

Would you like to work on this?
